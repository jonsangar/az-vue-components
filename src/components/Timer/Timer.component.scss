.timerModal{
  position: fixed;
  top: 0;
  left: 0;
  z-index: 100;
  width: 100vw;
  height: 100vh;

  /* Adding safe area  */
  padding-bottom: env(safe-area-inset-bottom);
  padding-bottom: var(--notch-inset-bottom);
  padding-left: env(safe-area-inset-left);
  padding-left: var(--notch-inset-left);
  background-color: rgba(0, 0, 0, 0.8);

  a {
    cursor: pointer;
  }

  table{
    tr:last-child{
      td{
        border: none;
      }
    }
  }

  &__backButton{
    position: absolute;
    top: simple();
    left: simple();
    cursor: pointer;
  }

  &__wrapper {
    position: absolute;
    top: 0 ;
    right: 0;
    bottom: 0;
    left: 0;
    width: calc(100vw - #{simple()});
    height: fit-content;
    padding: 0;
    margin: auto;
    overflow: hidden;
    background-color: color(gray8);
    border-radius: $border-radius ;

  }

  &__title {
    // @extend %base-bold-black;

    padding: simple() 0 simple();
    font-weight: bold;
    text-align: center;
  }

  &__message{
    padding: 0 double();
  }


  &__buttons {
    position: absolute;
    bottom: 0;
    display: flex;
    width: 100%;
    padding-top: simple();
  }

  &__button {
    @extend %base-normal-black;

    width: 100%;
    height: double();
    overflow: hidden;
    line-height: double();
    color: color(gray8) !important;
    text-align: center;
    cursor: pointer;
    border-top: 1px solid $border-color;

    &:not(:last-child) {
      border-right: 1px solid $border-color;
    }
  }

  &__button--brand{
    background-image: radial-gradient(circle at 0% 0%, color(brand-light), color(brand));
  }

  &__button--cancel{
    background-image: radial-gradient(circle at 0% 0%, color(gray6), color(gray4));
  }

  &__backButton{
    width: 20%;
    padding: 0 0 half();
  }

  h1, h2, h3, h4, h5, h6{
    &.colour-title{
      padding-bottom: half();
      margin-bottom: half();
      font-weight: bold;

      &.color-1{
        color: #00AAA1;

        &.underline{
          border-bottom: 2px solid #00AAA1;
        }
      }

      &.color-2{
        color: #FFB837;

        &.underline{
            border-bottom: 2px solid #FFB837;
        }
      }

      &.color-3{
        color: #F3975B;

        &.underline{
            border-bottom: 2px solid #F3975B;
        }
      }

      &.color-4{
        color: #D04F53;

        &.underline{
          border-bottom: 2px solid #D04F53;
        }
      }

      &.color-5{
        color: #965379;

        &.underline{
          border-bottom: 2px solid #965379;
        }
      }
    }
  }

  .fade-enter-active, .fade-leave-active {
    transition: opacity .5s
  }
  .fade-enter, .fade-leave-to /* .fade-leave-active below version 2.1.8 */ {
    opacity: 0
  }
}

@include media(landscape) {
  .timerModal{
    &__wrapper {
      max-width: 370px;
    }
  }
}
