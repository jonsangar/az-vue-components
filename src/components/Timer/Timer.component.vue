<template lang="pug">
.timerModal(
   @click="closeOnClick && close($event)")

  .timerModal__wrapper
    .timerModal__title.font-weight-bold.padding-half(
      v-if="title") {{title}}
    .timerModal__message.text-center(
      v-if="message") {{ message }}
    .picker-group.padding-half.margin-bottom-double
      scroll-picker.hours-picker(
        :options="hourOptions"
        v-model="hours")
      scroll-picker.minutes-picker(
        :options="minutesOptions"
        v-model="minutes")


    .timerModal__buttons.margin-top(v-if="buttons")
      .timerModal__button.timerModal__button--cancel(
        v-if="cancelBtn"
        @click="cancel") {{ getBtnText(cancelBtnText) }}
      .timerModal__button.timerModal__button--brand(
        v-if="acceptBtn"
            @click="accept") {{ getBtnText(acceptBtnText) }}
</template>

<script>
export default {
props: {
    value: {
      type: Number,
      required: true,
    },

    maxHours: {
      type: Number,
      default: 24,
      validator: value => value >= 0
    },

    maxMinutes: {
      type: Number,
      default: 59,
      validator: value => value >= 0 && value < 60
    },

    maxTime: {
      type: Number,
      default: undefined,
      validator: value => value >= 0 || undefined
    },

    title: {
      type: String,
      required: false
    },

    message: {
      type: String,
      required: false
    },

    buttons: {
      type: Boolean,
      default: false
    },

    backButton: {
      type: Boolean,
      default: false
    },

    acceptBtn: {
      type: Boolean,
      default: true
    },

    cancelBtn: {
      type: Boolean,
      default: true
    },

    cancelBtnText: {
      type: String,
    },

    acceptBtnText: {
      type: String,
    },

    textAlign: {
      type: String,
      default: 'center'
    },

    closeOnClick: {
      type: Boolean,
      default: true
    },
},

data() {
  return {
    localValue: 0,
    hours: 0,
    minutes: 0,
    hourOptions: [],
    minutesOptions: []
  }
},

watch: {
  value(newValue) {
    this.localValue = newValue
  },

  localHours(newValue) {
    this.$emit('input', newValue)
  },

  hours(value){
    if(this.maxTime) {
      const maxHours = Math.trunc(this.maxTime / 60);
      this.minutesOptions = [];
      if(value === maxHours) {
        const minutes = this.maxTime - maxHours * 60;
        for(let m = 0; m <= minutes; m++) {
          this.minutesOptions.push({
            name: `${m}m`,
            value: m,
          })
        }
      } else {
        for(let m = 0; m <= 59; m++) {
          this.minutesOptions.push({
            name: `${m}m`,
            value: m,
          })
        }
      }
    }
    this.localValue = this.parseToMinutes();
    this.$emit('input', this.localValue)
  },

  minutes(){
    this.localValue = this.parseToMinutes();
    this.$emit('input', this.localValue)
  }


},

methods: {

  parseToMinutes(){
    const hoursToMinutes = this.hours * 60;

    return hoursToMinutes + this.minutes;
  },

  getBtnText(btnText){
    if(typeof btnText === 'function'){
      return btnText();
    }

    return btnText;
  },

  accept(){

    if(this.callback) {
      console.log("Has callback");
      this.callback('confirm');
      this.clear();
    }


    this.$emit('accept', this.localValue);

  },

  close(e) {
    if(e.target !== e.currentTarget) return;

    if(this.callback) {
      this.callback('cancel');
      this.clear();
    }

    this.$emit('close');
  },

  cancel(e) {
    if(e.target !== e.currentTarget) return;

    if(this.callback) {
      this.callback('cancel');
      this.clear();
    }

    this.$emit('cancel');
  },

  showItemInfo(item) {
    this.$emit('showInfo',item);
  },

  initTimerValues() {
    if(this.maxTime !== undefined) {
      const hours = Math.trunc(this.maxTime / 60);
      const minutes = hours < 1 ? this.maxTime%60 : 59;
      for(let h = 0 ; h <= hours; h++) {
        this.hourOptions.push({
          name: `${h}h`,
          value: h,
        })
      }

      for(let m = 0; m <= minutes; m++) {
        this.minutesOptions.push({
          name: `${m}m`,
          value: m,
        })
      }
    } else {
      for(let h = 0 ; h <= this.maxHours; h++) {
        this.hourOptions.push({
          name: `${h}h`,
          value: h,
        })
      }

      for(let m = 0; m <= this.maxMinutes; m++) {
        this.minutesOptions.push({
          name: `${m}m`,
          value: m,
        })
      }
    }
  },

  initTimer() {
    this.localValue = this.value;
    this.hours = Math.trunc(this.value / 60);
    this.minutes = this.value%60
  }

},

created() {
  this.initTimerValues()
  this.initTimer();
}

}
</script>

<style lang="scss" src="./Timer.component.scss" scoped></style>
