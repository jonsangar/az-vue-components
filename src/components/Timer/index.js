import Timer from './Timer.component.vue';

export default Timer;
