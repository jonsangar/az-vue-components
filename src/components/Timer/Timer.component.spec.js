import { shallowMount, createLocalVue } from '@vue/test-utils';
import VueScrollPicker from 'vue-scroll-picker'
import Timer from './Timer.component.vue';


const localVue = createLocalVue();
localVue.use(VueScrollPicker);

describe('Timer.component.vue', () => {

  const build = () => {
    const wrapper = shallowMount(Timer, {
      localVue,
      propsData: {
        value: 120
      }
    })

    return {
      wrapper
    }
  }

  it('Renderizado del componente con snapshot', () => {
    const { wrapper } = build();

    wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Comprobar que el valor de los scroll pickers se parsea correctamente', () => {
    const { wrapper } = build();

    wrapper.vm.$nextTick();
    // Por defecto value es 120 min. Equivale a 2h y 0m
    const hoursPicker = wrapper.find('.hours-picker');
    const minutesPicker = wrapper.find('.minutes-picker');

    expect(hoursPicker.attributes().value).toBe("2");
    expect(minutesPicker.attributes().value).toBe("0");

  });

  it('Comprobar que no se muestra titulo cuando no se recibe', () => {
    const {wrapper} = build();

    wrapper.vm.$nextTick();

    const title = wrapper.find('.timerModal__title');

    expect(title.exists()).toBe(false)
  });

  it('Comprobar que se muestra titulo cuando se recibe la prop', () => {
    const wrapper = shallowMount(Timer, {
      localVue,
      propsData: {
        value: 120,
        title: 'Temporizador'
      }
    })

    wrapper.vm.$nextTick();

    const title = wrapper.find('.timerModal__title');

    expect(wrapper.html()).toMatchSnapshot();

    expect(title.isVisible()).toBe(true);
    expect(title.text()).toStrictEqual('Temporizador');


  });

  it('Comprobar que no se muestra message cuando no se recibe', () => {
    const {wrapper} = build();

    wrapper.vm.$nextTick();

    const message = wrapper.find('.timerModal__message');

    expect(message.exists()).toBe(false)
  });

  it('Comprobar que se muestra message cuando se recibe la prop', () => {
    const wrapper = shallowMount(Timer, {
      localVue,
      propsData: {
        value: 120,
        message: 'Lorem ipsum dolor sit amet'
      }
    })

    wrapper.vm.$nextTick();

    const message = wrapper.find('.timerModal__message');

    expect(wrapper.html()).toMatchSnapshot();

    expect(message.isVisible()).toBe(true);
    expect(message.text()).toStrictEqual('Lorem ipsum dolor sit amet');

  });

  it('Comprobar que no se muestran botones de confirmación si no se recibe', () => {
    const {wrapper} = build();

    wrapper.vm.$nextTick();

    const buttons = wrapper.find('.timerModal__buttons');

    expect(buttons.exists()).toBe(false)
  });

  it('Comprobar que se muestran botones de confirmación si se recibe la prop', () => {
    const wrapper = shallowMount(Timer, {
      localVue,
      propsData: {
        value: 120,
        buttons: true
      }
    });

    wrapper.vm.$nextTick();

    const buttons = wrapper.find('.timerModal__buttons');

    expect(wrapper.html()).toMatchSnapshot();

    expect(buttons.isVisible()).toBe(true);
  })
});
