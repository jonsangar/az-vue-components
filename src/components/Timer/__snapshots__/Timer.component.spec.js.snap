// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer.component.vue Comprobar que se muestra message cuando se recibe la prop 1`] = `
<div class="timerModal">
  <div class="timerModal__wrapper">
    <!---->
    <div class="timerModal__message text-center">Lorem ipsum dolor sit amet</div>
    <div class="picker-group padding-half margin-bottom-double">
      <scroll-picker-stub value="2" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="hours-picker"></scroll-picker-stub>
      <scroll-picker-stub value="0" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="minutes-picker"></scroll-picker-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Timer.component.vue Comprobar que se muestra titulo cuando se recibe la prop 1`] = `
<div class="timerModal">
  <div class="timerModal__wrapper">
    <div class="timerModal__title font-weight-bold padding-half">Temporizador</div>
    <!---->
    <div class="picker-group padding-half margin-bottom-double">
      <scroll-picker-stub value="2" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="hours-picker"></scroll-picker-stub>
      <scroll-picker-stub value="0" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="minutes-picker"></scroll-picker-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Timer.component.vue Comprobar que se muestran botones de confirmaci√≥n si se recibe la prop 1`] = `
<div class="timerModal">
  <div class="timerModal__wrapper">
    <!---->
    <!---->
    <div class="picker-group padding-half margin-bottom-double">
      <scroll-picker-stub value="2" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="hours-picker"></scroll-picker-stub>
      <scroll-picker-stub value="0" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="minutes-picker"></scroll-picker-stub>
    </div>
    <div class="timerModal__buttons margin-top">
      <div class="timerModal__button timerModal__button--cancel"></div>
      <div class="timerModal__button timerModal__button--brand"></div>
    </div>
  </div>
</div>
`;

exports[`Timer.component.vue Renderizado del componente con snapshot 1`] = `
<div class="timerModal">
  <div class="timerModal__wrapper">
    <!---->
    <!---->
    <div class="picker-group padding-half margin-bottom-double">
      <scroll-picker-stub value="2" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="hours-picker"></scroll-picker-stub>
      <scroll-picker-stub value="0" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" dragsensitivity="1.7" touchsensitivity="1.7" scrollsensitivity="1" empty="No Items" class="minutes-picker"></scroll-picker-stub>
    </div>
    <!---->
  </div>
</div>
`;
